odl/util/numerics.py

```diff
@@ -297,6 +303,593 @@ def fast_1d_tensor_mult(ndarr, onedim_arrs, axes=None, out=None):
+
+    # Handle frac_left and num_left parameters
+    if frac_left is not None and num_left is not None:
+        raise ValueError('the options `frac_left` and `num_left` cannot be '
+                         'combined')
+    if num_left is None:
+        # Compute num_left from frac_left
+        if frac_left is None:
+            frac_left = 0.5
+        frac_left, frac_left_in = normalized_scalar_param_list(
+            frac_left, out.ndim, param_conv=float, keep_none=False,
+            return_nonconv=True)
+        num_left = [round(frac_l * abs(n_orig - n_new))
+                    for frac_l, n_orig, n_new in zip(frac_left, arr.shape,
+                                                     out.shape)]
+        for i, (fl, fl_in) in enumerate(zip(frac_left, frac_left_in)):
+            if not 0.0 <= fl <= 1.0:
+                raise ValueError('in axis {}: `frac_left` must lie between 0 '
+                                 'and 1, got {}'.format(i, fl_in))
+    else:
+        num_left = normalized_scalar_param_list(
+            num_left, out.ndim, param_conv=safe_int_conv, keep_none=False)
+
+    # Handle padding
+    pad_mode, pad_mode_in = str(pad_mode).lower(), pad_mode
+    if pad_mode not in _SUPPORTED_PAD_MODES:
+        raise ValueError("`pad_mode` '{}' not understood".format(pad_mode_in))
+
+    if (pad_mode == 'constant' and
+        not np.can_cast(pad_const, out.dtype) and
+        any(n_new > n_orig
+            for n_orig, n_new in zip(arr.shape, out.shape))):
+        raise ValueError('`pad_const` {} cannot be safely cast to the data '
+                         'type {} of the output array'
+                         ''.format(pad_const, out.dtype))
+
+    # Direction
+    direction, direction_in = str(direction).lower(), direction
+    if direction not in ('forward', 'adjoint'):
+        raise ValueError("`direction` '{}' not understood"
+                         "".format(direction_in))
+
+    # Calculate the slices for the inner and outer parts
+    arr_slc, out_slc, pad_l_slc, pad_r_slc = [], [], [], []
+    for i, (n_orig, n_new, num_l) in enumerate(zip(arr.shape, out.shape,
+                                                   num_left)):
+        if n_new < n_orig:
+            # Simple case: remove according to num_left
+            n_remove = n_orig - n_new
+            istart = num_l
+            istop = n_orig - (n_remove - istart)
+            arr_slc.append(slice(istart, istop))
+            out_slc.append(slice(None))
+
+            # Make trivial entries for padding slice lists
+            pad_l_slc.append(slice(0))
+            pad_r_slc.append(slice(0))
+
+        elif n_new > n_orig:
+            # Padding case: calculate start and stop indices for the
+            # bigger new array, together with the padding slices
+            n_add = n_new - n_orig
+            istart = num_l
+            istop = n_new - (n_add - istart)
+
+            n_pad_l = len(range(istart))
```

Comment:
Wut? Why not simply `n_pad_l = istart`? Or if pad can be negative, `n_pad_l = max(0, istart)`